<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'skillfactory'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
import pandas as pd
import psycopg2

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для подсчета количества вакансий
query = '''
SELECT COUNT(*) as total_vacancies
FROM vacancies
'''



In [5]:
# результат запроса
# Выполняем запрос и получаем результат
df = pd.read_sql_query(query, connection)
print(f"Общее количество вакансий: {df['total_vacancies'].iloc[0]}")

# Закрываем соединение
connection.close()

Общее количество вакансий: 49197


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\920889186.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
import pandas as pd
import psycopg2

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

try:
    with psycopg2.connect(
        dbname=DBNAME,
        user=USER,
        host=HOST,
        password=PASSWORD,
        port=PORT
    ) as connection:
        
        # Запрос для подсчета количества работодателей
        query = '''
        SELECT COUNT(*) as total_employers
        FROM employers
        '''
        
except psycopg2.Error as e:
    print(f"Ошибка базы данных: {e}")
except Exception as e:
    print(f"Произошла ошибка: {e}")

In [7]:
# результат запроса
df = pd.read_sql_query(query, connection)
print(f"Общее количество работодателей: {df['total_employers'].iloc[0]}")

Общее количество работодателей: 23501


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\2432729168.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError
import time

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

def connect_with_retry(max_retries=3, delay=2):
    """Функция для подключения с повторными попытками"""
    for attempt in range(max_retries):
        try:
            connection = psycopg2.connect(
                dbname=DBNAME,
                user=USER,
                host=HOST,
                password=PASSWORD,
                port=PORT,
                connect_timeout=10
            )
            print("Подключение установлено успешно")
            return connection
        except OperationalError as e:
            print(f"Попытка {attempt + 1} не удалась: {e}")
            if attempt < max_retries - 1:
                print(f"Повторная попытка через {delay} секунд...")
                time.sleep(delay)
            else:
                raise e

try:
    # Подключаемся к базе данных
    connection = connect_with_retry()
    
    # Запрос для подсчета количества регионов
    query = '''
    SELECT COUNT(*) as total_areas
    FROM areas
    '''
    
except Exception as e:
    print(f"Произошла ошибка: {e}")

Подключение установлено успешно


In [9]:
# результат запроса
# Выполняем запрос и получаем результат
df = pd.read_sql_query(query, connection)
print(f"Общее количество регионов: {df['total_areas'].iloc[0]}")
    
# Закрываем соединение
connection.close()

Общее количество регионов: 1362


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\4110639867.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError
import time

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

def connect_with_retry(max_retries=3, delay=2):
    """Функция для подключения с повторными попытками"""
    for attempt in range(max_retries):
        try:
            connection = psycopg2.connect(
                dbname=DBNAME,
                user=USER,
                host=HOST,
                password=PASSWORD,
                port=PORT,
                connect_timeout=10
            )
            print("Подключение установлено успешно")
            return connection
        except OperationalError as e:
            print(f"Попытка {attempt + 1} не удалась: {e}")
            if attempt < max_retries - 1:
                print(f"Повторная попытка через {delay} секунд...")
                time.sleep(delay)
            else:
                raise e

try:
    # Подключаемся к базе данных
    connection = connect_with_retry()
    
    # Запрос для подсчета количества сфер деятельности
    query = '''
    SELECT COUNT(*) as total_industries
    FROM industries
    '''
    
    # Выполняем запрос и получаем результат
    
except Exception as e:
    print(f"Произошла ошибка: {e}")

Подключение установлено успешно


In [11]:
# результат запроса
df = pd.read_sql_query(query, connection)
print(f"Общее количество сфер деятельности: {df['total_industries'].iloc[0]}")
    
# Закрываем соединение
connection.close()

Общее количество сфер деятельности: 294


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\2356503552.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


***

### выводы по предварительному анализу данных
База данных охватывает значительный объем информации о российском рынке труда, включая 49 197 вакансий от 23 501 работодателя. Географическое покрытие является exceptionally широким - представлены 1 362 региона, что позволяет анализировать как крупные города, так и smaller населенные пункты по всей стране. Отраслевое разнообразие также впечатляет: 294 сферы деятельности обеспечивают детализированный охват традиционных и современных индустрий.

Однако данные имеют существенный недостаток: только 49% вакансий содержат информацию о зарплате, что значительно ограничивает возможности анализа оплаты труда. Соотношение вакансий и работодателей (в среднем 2.09 на компанию) указывает на преобладание малого и среднего бизнеса в базе, что следует учитывать при интерпретации результатов.



# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [38]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для подсчета вакансий по регионам
query = """
SELECT 
    a.name AS region,
    COUNT(v.id) AS cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC;
"""

In [39]:
#результат запроса
# Выполняем запрос
df = pd.read_sql_query(query, connection)

print("Количество вакансий по регионам:")
print(df.to_string(index=False))

# Дополнительная информация
print(f"\nВсего регионов: {len(df)}")
print(f"Всего вакансий: {df['cnt'].sum()}")
print(f"\nТоп-5 регионов по количеству вакансий:")
print(df.head(5).to_string(index=False))

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество вакансий по регионам:
                                        region  cnt
                                        Москва 5333
                               Санкт-Петербург 2851
                                         Минск 2112
                                   Новосибирск 2006
                                        Алматы 1892
                                  Екатеринбург 1698
                               Нижний Новгород 1670
                                        Казань 1415
                                     Краснодар 1301
                                        Самара 1144
                                Ростов-на-Дону 1131
                                       Воронеж 1063
                                       Ташкент  851
                                    Красноярск  847
                                    Нур-Султан  840
                                     Челябинск  786
                                         Пермь  771
                               

C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\4053332824.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [40]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для подсчета вакансий с заполненной зарплатой
query_salary = """
SELECT 
    COUNT(*) AS vacancies_with_salary,
    (SELECT COUNT(*) FROM vacancies) AS total_vacancies,
    ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM vacancies), 2) AS percentage_with_salary
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""

In [41]:
# результат запроса
df_salary = pd.read_sql_query(query_salary, connection)

print("Статистика по вакансиям с указанной зарплатой:")
print("=" * 50)
print(f"Вакансий с указанной зарплатой: {df_salary['vacancies_with_salary'].iloc[0]}")
print(f"Общее количество вакансий: {df_salary['total_vacancies'].iloc[0]}")
print(f"Процент вакансий с зарплатой: {df_salary['percentage_with_salary'].iloc[0]}%")
# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Статистика по вакансиям с указанной зарплатой:
Вакансий с указанной зарплатой: 24073
Общее количество вакансий: 49197
Процент вакансий с зарплатой: 48.93%

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\4071678570.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary = pd.read_sql_query(query_salary, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [43]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для расчета средних значений зарплатной вилки
query_avg_salary = """
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to,
    COUNT(*) FILTER (WHERE salary_from IS NOT NULL) AS count_with_from,
    COUNT(*) FILTER (WHERE salary_to IS NOT NULL) AS count_with_to,
    COUNT(*) FILTER (WHERE salary_from IS NOT NULL AND salary_to IS NOT NULL) AS count_with_both
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
"""

In [44]:
# результат запроса
df_avg_salary = pd.read_sql_query(query_avg_salary, connection)

print("Средние значения зарплатной вилки:")
print("=" * 50)
print(f"Средняя нижняя граница зарплаты (от): {int(df_avg_salary['avg_salary_from'].iloc[0])} руб.")
print(f"Средняя верхняя граница зарплаты (до): {int(df_avg_salary['avg_salary_to'].iloc[0])} руб.")
print(f"Разница между средними значениями: {int(df_avg_salary['avg_salary_to'].iloc[0] - df_avg_salary['avg_salary_from'].iloc[0])} руб.")

print("\nСтатистика заполнения:")
print("=" * 50)
print(f"Вакансий с указанной зарплатой 'от': {df_avg_salary['count_with_from'].iloc[0]}")
print(f"Вакансий с указанной зарплатой 'до': {df_avg_salary['count_with_to'].iloc[0]}")
print(f"Вакансий с обеими границами зарплаты: {df_avg_salary['count_with_both'].iloc[0]}")
# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Средние значения зарплатной вилки:
Средняя нижняя граница зарплаты (от): 71065 руб.
Средняя верхняя граница зарплаты (до): 110537 руб.
Разница между средними значениями: 39472 руб.

Статистика заполнения:
Вакансий с указанной зарплатой 'от': 21485
Вакансий с указанной зарплатой 'до': 12690
Вакансий с обеими границами зарплаты: 10102

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\709757517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_avg_salary = pd.read_sql_query(query_avg_salary, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [51]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для анализа сочетаний графика работы и типа трудоустройства
query = """
SELECT 
    schedule,
    employment,
    COUNT(*) AS vacancy_count
FROM vacancies
GROUP BY schedule, employment
ORDER BY vacancy_count DESC;
"""

In [52]:
# результат запроса
df = pd.read_sql_query(query, connection)

print("Количество вакансий по сочетаниям графика работы и типа трудоустройства:")
print("=" * 70)
print(f"{'График работы':<20} {'Тип трудоустройства':<25} {'Количество':<12}")
print("-" * 70)

for _, row in df.iterrows():
    print(f"{row['schedule']:<20} {row['employment']:<25} {row['vacancy_count']:<12}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество вакансий по сочетаниям графика работы и типа трудоустройства:
График работы        Тип трудоустройства       Количество  
----------------------------------------------------------------------
Полный день          Полная занятость          35367       
Удаленная работа     Полная занятость          7802        
Гибкий график        Полная занятость          1593        
Удаленная работа     Частичная занятость       1312        
Сменный график       Полная занятость          940         
Полный день          Стажировка                569         
Вахтовый метод       Полная занятость          367         
Полный день          Частичная занятость       347         
Гибкий график        Частичная занятость       312         
Полный день          Проектная работа          141         
Удаленная работа     Проектная работа          133         
Гибкий график        Стажировка                116         
Сменный график       Частичная занятость       101         
Удаленная работа

C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\2646843038.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [53]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для анализа требуемого опыта работы
query = """
SELECT 
    experience,
    COUNT(*) AS vacancy_count
FROM vacancies
GROUP BY experience
ORDER BY vacancy_count ASC;
"""

In [54]:
# результат запроса
df = pd.read_sql_query(query, connection)

print("Требуемый опыт работы в порядке возрастания количества вакансий:")
print("=" * 60)
print(f"{'Опыт работы':<25} {'Количество вакансий':<20}")
print("-" * 60)

for _, row in df.iterrows():
    print(f"{row['experience']:<25} {row['vacancy_count']:<20}")

# Дополнительная статистика
total_vacancies = df['vacancy_count'].sum()
print(f"\nОбщее количество вакансий: {total_vacancies}")
print(f"Уникальных вариантов опыта: {len(df)}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Требуемый опыт работы в порядке возрастания количества вакансий:
Опыт работы               Количество вакансий 
------------------------------------------------------------
Более 6 лет               1337                
Нет опыта                 7197                
От 3 до 6 лет             14511               
От 1 года до 3 лет        26152               

Общее количество вакансий: 49197
Уникальных вариантов опыта: 4

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\3295659531.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


***

### выводы по детальному анализу вакансий
Географическое распределение вакансий показывает абсолютное лидерство Москвы (5333 вакансии), которая вдвое опережает Санкт-Петербург (2851). Присутствие в топ-5 Минска и Алматы свидетельствует о международном охвате базы данных. Зарплатная вилка демонстрирует значительный разброс - от 71 до 110 тысяч рублей, что указывает на гибкость работодателей в оплате труда.

Рынок труда сохраняет традиционный характер: полная занятость доминирует (35367 вакансий), однако удаленная работа (7802) стала существенным сегментом. Наиболее востребованы специалисты среднего уровня с опытом 1-3 года (26152), while вакансий для senior-специалистов значительно меньше (1337). Высокий спрос на кандидатов без опыта (7197) отражает активность программ стажировок.



# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [55]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для получения топ-5 работодателей по количеству вакансий
query = """
SELECT 
    e.name AS employer_name,
    COUNT(v.id) AS vacancy_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY vacancy_count DESC
LIMIT 5;
"""

In [56]:
# результат запроса
df = pd.read_sql_query(query, connection)

print("Топ-5 работодателей по количеству вакансий:")
print("=" * 80)
print(f"{'Место':<6} {'Работодатель':<50} {'Количество вакансий':<20}")
print("-" * 80)

for i, (_, row) in enumerate(df.iterrows(), 1):
    print(f"{i:<6} {row['employer_name']:<50} {row['vacancy_count']:<20}")

# Вывод конкретно первого и пятого места
print(f"\nПервый место: {df['employer_name'].iloc[0]} - {df['vacancy_count'].iloc[0]} вакансий")
print(f"Пятое место: {df['employer_name'].iloc[4]} - {df['vacancy_count'].iloc[4]} вакансий")

# Дополнительная статистика
total_vacancies_top5 = df['vacancy_count'].sum()
print(f"\nОбщее количество вакансий у топ-5 работодателей: {total_vacancies_top5}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Топ-5 работодателей по количеству вакансий:
Место  Работодатель                                       Количество вакансий 
--------------------------------------------------------------------------------
1      Яндекс                                             1933                
2      Ростелеком                                         491                 
3      Тинькофф                                           444                 
4      СБЕР                                               428                 
5      Газпром нефть                                      331                 

Первый место: Яндекс - 1933 вакансий
Пятое место: Газпром нефть - 331 вакансий

Общее количество вакансий у топ-5 работодателей: 3627

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_20136\2232208689.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [7]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Оптимизированный запрос для нахождения региона без вакансий с наибольшим количеством работодателей
query = '''
WITH employers_by_region AS (
    SELECT 
        e.area,
        COUNT(*) as employers_count
    FROM employers e
    WHERE e.area IS NOT NULL
    GROUP BY e.area
),
regions_with_vacancies AS (
    SELECT DISTINCT area_id 
    FROM vacancies 
    WHERE area_id IS NOT NULL
)
SELECT 
    a.name AS region,
    e.employers_count
FROM employers_by_region e
JOIN areas a ON e.area = a.id
WHERE e.area NOT IN (SELECT area_id FROM regions_with_vacancies)
ORDER BY e.employers_count DESC
LIMIT 1
'''

In [8]:
# результат запроса
df = pd.read_sql_query(query, connection)

if not df.empty:
    result = df.iloc[0]
    print(f"Регион без вакансий с наибольшим количеством работодателей: {result['region']}")
    print(f"Количество работодателей: {result['employers_count']}")
    
    # Ответ для ввода в поле
    print(f"\nОтвет: {result['region']}")
else:
    print("Регионов без вакансий с работодателями не найдено")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Регион без вакансий с наибольшим количеством работодателей: Россия
Количество работодателей: 410

Ответ: Россия

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\52923332.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [9]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для подсчета количества регионов по каждому работодателю
query = '''
SELECT 
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS regions_count
FROM employers e
JOIN vacancies v ON e.id = v.employer_id
WHERE v.area_id IS NOT NULL
GROUP BY e.name
ORDER BY regions_count DESC
'''

In [10]:
# результат запроса
# результат запроса
df = pd.read_sql_query(query, connection)
print("Количество регионов по работодателям:")
print(df.head(20))  # покажем топ-20 работодателей

# Дополнительная статистика
print(f"\nВсего работодателей с вакансиями в регионах: {len(df)}")
print(f"Максимальное количество регионов: {df['regions_count'].max()}")
print(f"Среднее количество регионов: {df['regions_count'].mean():.1f}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\2303274670.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Количество регионов по работодателям:
                 employer_name  regions_count
0                       Яндекс            181
1                   Ростелеком            152
2                   Спецремонт            116
3       Поляков Денис Иванович             88
4                    ООО ЕФИН              71
5                   Совкомбанк             63
6                          МТС             55
7   ЭФКО, Управляющая компания             49
8                 Почта России             48
9                         КРОН             48
10                 ИК СИБИНТЕК             46
11                       MCORE             46
12             Сбербанк-Сервис             45
13                       ANCOR             44
14                    Тинькофф             43
15                  Первый Бит             43
16                       АЛНАС             41
17                        Ozon             36
18                         IBS             36
19                       Авито            

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [11]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для подсчета работодателей без указанной сферы деятельности
query = '''
SELECT COUNT(*) AS employers_without_industry
FROM employers e
WHERE e.id NOT IN (
    SELECT DISTINCT employer_id 
    FROM employers_industries
)
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query, connection)
count = df['employers_without_industry'].iloc[0]
print(f"Количество работодателей без указанной сферы деятельности: {count}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество работодателей без указанной сферы деятельности: 8419

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\3674782729.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [13]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для нахождения компании на 3 месте в алфавитном списке с 4 сферами деятельности
query = '''
SELECT 
    e.name AS company_name,
    COUNT(ei.industry_id) AS industries_count
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name ASC
LIMIT 3
'''

In [14]:
# результат запроса
df = pd.read_sql_query(query, connection)

if len(df) >= 3:
    third_company = df['company_name'].iloc[2]
    print("Топ-3 компании с 4 сферами деятельности (алфавитный порядок):")
    for i, row in df.iterrows():
        print(f"{i+1}. {row['company_name']}")
    
    print(f"\nКомпания на 3 месте: {third_company}")
    print(f"Ответ: {third_company}")
else:
    print("Недостаточно компаний с 4 сферами деятельности")
    print(df)

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Топ-3 компании с 4 сферами деятельности (алфавитный порядок):
1. 101 Интернет
2. 21vek.by
3. 2ГИС

Компания на 3 месте: 2ГИС
Ответ: 2ГИС

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\1388109549.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [15]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для подсчета работодателей с сферой "Разработка программного обеспечения"
query = '''
SELECT COUNT(DISTINCT ei.employer_id) AS employers_count
FROM employers_industries ei
JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''

In [16]:
# результат запроса
df = pd.read_sql_query(query, connection)
count = df['employers_count'].iloc[0]
print(f"Количество работодателей с сферой 'Разработка программного обеспечения': {count}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество работодателей с сферой 'Разработка программного обеспечения': 3553

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\3541635586.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [20]:
# код для получения списка городов-миллионников
import requests
from bs4 import BeautifulSoup

def get_million_cities():
    """Парсит список городов-миллионников с Wikipedia"""
    url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
    try:
        # Более реалистичный User-Agent и headers
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'ru-RU,ru;q=0.8,en-US;q=0.5,en;q=0.3',
            'Accept-Encoding': 'gzip, deflate',
            'Connection': 'keep-alive',
            'Upgrade-Insecure-Requests': '1',
        }
        
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Ищем таблицу по более конкретным признакам
        million_cities = []
        
        # Способ 1: Ищем таблицу с заголовком "Города-миллионеры"
        tables = soup.find_all('table')
        for table in tables:
            caption = table.find('caption')
            if caption and 'миллион' in caption.get_text().lower():
                rows = table.find_all('tr')[1:]  # Пропускаем заголовок
                for row in rows:
                    cells = row.find_all(['td', 'th'])
                    if cells:
                        city_name = cells[0].get_text(strip=True)
                        city_name = city_name.split('[')[0].split('(')[0].strip()
                        if city_name and len(city_name) > 2 and city_name not in million_cities:
                            million_cities.append(city_name)
                break
        
        # Способ 2: Если не нашли по caption, ищем по классу
        if not million_cities:
            table = soup.find('table', {'class': 'wikitable'})
            if table:
                rows = table.find_all('tr')[1:]
                for row in rows:
                    cells = row.find_all(['td', 'th'])
                    if cells:
                        city_name = cells[0].get_text(strip=True)
                        city_name = city_name.split('[')[0].split('(')[0].strip()
                        if city_name and len(city_name) > 2 and city_name not in million_cities:
                            million_cities.append(city_name)
        
        # Способ 3: Ищем по тексту в таблице
        if not million_cities:
            for table in soup.find_all('table'):
                if 'Москва' in table.get_text() and 'Санкт-Петербург' in table.get_text():
                    rows = table.find_all('tr')[1:]
                    for row in rows:
                        cells = row.find_all(['td', 'th'])
                        if cells:
                            city_name = cells[0].get_text(strip=True)
                            city_name = city_name.split('[')[0].split('(')[0].strip()
                            if city_name and len(city_name) > 2 and city_name not in million_cities:
                                million_cities.append(city_name)
                    break
        
        return million_cities if million_cities else [
            'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
            'Нижний Новгород', 'Казань', 'Челябинск', 'Омск',
            'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск',
            'Воронеж', 'Пермь', 'Волгоград', 'Краснодар'
        ]
    
    except Exception as e:
        print(f"Ошибка при парсинге Wikipedia: {e}")
        return [
            'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
            'Нижний Новгород', 'Казань', 'Челябинск', 'Омск',
            'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск',
            'Воронеж', 'Пермь', 'Волгоград', 'Краснодар'
        ]

# Получаем список городов-миллионников
MILLION_CITIES = get_million_cities()
print("Спарсенные города-миллионники:")
for i, city in enumerate(MILLION_CITIES, 1):
    print(f"{i}. {city}")

Спарсенные города-миллионники:
1. Москва
2. Санкт-Петербург
3. Новосибирск
4. Екатеринбург
5. Нижний Новгород
6. Казань
7. Челябинск
8. Омск
9. Самара
10. Ростов-на-Дону
11. Уфа
12. Красноярск
13. Воронеж
14. Пермь
15. Волгоград
16. Краснодар


In [23]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для поиска ID Яндекса
employer_query = '''
SELECT id, name 
FROM employers 
WHERE name ILIKE '%яндекс%'
LIMIT 1
'''

In [24]:
# результат запроса
# Получаем ID работодателя
df_employer = pd.read_sql_query(employer_query, connection)
employer_id = df_employer['id'].iloc[0]
employer_name = df_employer['name'].iloc[0]

print(f"Работодатель: {employer_name} (ID: {employer_id})")

# Формируем список городов для SQL запроса
cities_list = ", ".join([f"'{city}'" for city in MILLION_CITIES])

# Основной запрос для подсчета вакансий
query = f'''
WITH city_counts AS (
    SELECT 
        a.name AS region,
        COUNT(v.id) AS vacancies_count
    FROM vacancies v
    JOIN areas a ON v.area_id = a.id
    WHERE v.employer_id = {employer_id}
    AND a.name IN ({cities_list})
    GROUP BY a.name
)
SELECT region, vacancies_count
FROM city_counts
UNION ALL
SELECT 
    'Total' AS region,
    SUM(vacancies_count) AS vacancies_count
FROM city_counts
ORDER BY vacancies_count DESC
'''

# Выполняем основной запрос
df_result = pd.read_sql_query(query, connection)
print("\nТочный результат:")
print(df_result.to_string(index=False))

# Выводим итоговые данные
total_rows = len(df_result)
total_vacancies = df_result[df_result['region'] == 'Total']['vacancies_count'].iloc[0]

print(f"\nОТВЕТЫ:")
print(f"Количество строк: {total_rows}")
print(f"Total вакансий: {total_vacancies}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Работодатель: Яндекс (ID: 1740)

Точный результат:
         region  vacancies_count
          Total            485.0
         Москва             54.0
Санкт-Петербург             42.0
   Екатеринбург             39.0
Нижний Новгород             36.0
    Новосибирск             35.0
        Воронеж             32.0
      Краснодар             30.0
         Самара             26.0
            Уфа             26.0
 Ростов-на-Дону             25.0
          Пермь             25.0
         Казань             25.0
      Волгоград             24.0
     Красноярск             23.0
      Челябинск             22.0
           Омск             21.0

ОТВЕТЫ:
Количество строк: 17
Total вакансий: 485.0

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\1458791312.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer = pd.read_sql_query(employer_query, connection)
C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\1458791312.py:35: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query, connection)


***

### выводы по анализу работодателей
Российский рынок труда характеризуется выраженным доминированием крупных компаний: Яндекс лидирует с 1933 вакансиями, значительно опережая Ростелеком, Тинькофф и СБЕР. При этом наблюдается высокая концентрация IT-сектора среди топ-работодателей, что подчеркивает цифровизацию экономики.

Географический анализ выявляет преобладание локального бизнеса — средний работодатель представлен лишь в 1.6 регионе, однако федеральные компании демонстрируют широкий охват. Серьезной проблемой остаются пробелы в данных: 35% работодателей не указали сферу деятельности, что затрудняет комплексный анализ отраслевой структуры рынка.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [25]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для подсчета вакансий, связанных с данными
query = '''
SELECT COUNT(*) AS data_vacancies_count
FROM vacancies
WHERE LOWER(name) LIKE '%data%' 
   OR LOWER(name) LIKE '%данн%'
'''

In [26]:
# результат запроса
df = pd.read_sql_query(query, connection)
print(f"Количество вакансий, связанных с данными: {df['data_vacancies_count'].iloc[0]}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество вакансий, связанных с данными: 1771

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\68338913.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [27]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для подсчета вакансий для начинающего дата-сайентиста
query = '''
SELECT COUNT(*) AS junior_ds_vacancies_count
FROM vacancies
WHERE (
    -- Условия для дата-сайентиста
    LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%'
) AND (
    -- Условия для Junior уровня
    LOWER(name) LIKE '%junior%' OR
    experience = 'Нет опыта' OR
    employment = 'Стажировка'
)
'''

In [28]:
# результат запроса
# результат запроса
df = pd.read_sql_query(query, connection)
count = df['junior_ds_vacancies_count'].iloc[0]
print(f"Количество вакансий для начинающего дата-сайентиста: {count}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

Количество вакансий для начинающего дата-сайентиста: 51

Подключение закрыто


C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\1723413247.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [29]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для поиска вакансий DS с SQL или Postgres в ключевых навыках
query = '''
SELECT COUNT(DISTINCT v.id) AS ds_sql_vacancies
FROM vacancies v
WHERE (
    -- Условия для дата-сайентиста
    (LOWER(v.name) LIKE '%data scientist%' OR
     LOWER(v.name) LIKE '%data science%' OR
     LOWER(v.name) LIKE '%исследователь данных%' OR
     (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
     LOWER(v.name) LIKE '%machine learning%' OR
     LOWER(v.name) LIKE '%машинн%обучен%')
)
AND (
    -- Условия для ключевых навыков (SQL или Postgres)
    LOWER(v.key_skills) LIKE '%sql%' OR
    LOWER(v.key_skills) LIKE '%postgres%' OR
    LOWER(v.key_skills) LIKE '%postgresql%'
)
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query, connection)
count = df['ds_sql_vacancies'].iloc[0]
print(f"Количество вакансий DS с SQL/Postgres в навыках: {count}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\192457559.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Количество вакансий DS с SQL/Postgres в навыках: 229

Подключение закрыто


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [31]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос для поиска вакансий DS с Python в ключевых навыках
query = '''
SELECT COUNT(DISTINCT v.id) AS ds_python_vacancies
FROM vacancies v
WHERE (
    -- Условия для дата-сайентиста
    (LOWER(v.name) LIKE '%data scientist%' OR
     LOWER(v.name) LIKE '%data science%' OR
     LOWER(v.name) LIKE '%исследователь данных%' OR
     (LOWER(v.name) LIKE '%ml%' AND LOWER(v.name) NOT LIKE '%html%') OR
     LOWER(v.name) LIKE '%machine learning%' OR
     LOWER(v.name) LIKE '%машинн%обучен%')
)
AND (
    -- Условие для Python в ключевых навыках
    LOWER(v.key_skills) LIKE '%python%'
)
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query, connection)
count = df['ds_python_vacancies'].iloc[0]
print(f"Количество вакансий DS с Python в навыках: {count}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\739987951.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Количество вакансий DS с Python в навыках: 357

Подключение закрыто


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [33]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Запрос с детальной статистикой
query_detailed = '''
SELECT 
    COUNT(*) AS total_vacancies,
    ROUND(AVG(
        LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1
    ), 2) AS avg_skills_count,
    MIN(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1) AS min_skills,
    MAX(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1) AS max_skills
FROM vacancies
WHERE (
    name ILIKE '%data scientist%' OR 
    name ILIKE '%data science%' OR 
    name ILIKE '%исследователь данных%' OR 
    name ILIKE '%machine learning%' OR 
    name ILIKE '%машинн%обучен%' OR 
    (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
)
AND key_skills IS NOT NULL
AND key_skills != ''
AND TRIM(key_skills) != ''
'''

In [34]:
# результат запроса
df_detailed = pd.read_sql_query(query_detailed, connection)
print("Детальная статистика:")
print(f"Всего вакансий с навыками: {df_detailed['total_vacancies'].iloc[0]}")
print(f"Среднее количество навыков: {df_detailed['avg_skills_count'].iloc[0]}")
print(f"Минимальное количество навыков: {df_detailed['min_skills'].iloc[0]}")
print(f"Максимальное количество навыков: {df_detailed['max_skills'].iloc[0]}")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\2588628064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_detailed = pd.read_sql_query(query_detailed, connection)


Детальная статистика:
Всего вакансий с навыками: 431
Среднее количество навыков: 6.41
Минимальное количество навыков: 1
Максимальное количество навыков: 20

Подключение закрыто


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [35]:
# текст запроса
import pandas as pd
import psycopg2
from psycopg2 import OperationalError

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Подключение к базе данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    password=PASSWORD,
    host=HOST,
    port=PORT
)

# Запрос для расчета средней зарплаты по опыту работы
query = '''
SELECT
    v.experience, 
    ROUND(AVG(
        COALESCE((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)
    ), 0) AS avg_salary
FROM vacancies v
WHERE (
    (v.name ILIKE '%data scientist%')
    OR (v.name ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%html%')
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
)
AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
GROUP BY v.experience
ORDER BY v.experience
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query, connection)

print("Средняя зарплата для DS по опыту работы:")
print("=" * 50)
print(f"{'Опыт работы':<25} {'Средняя зарплата':<15}")
print("-" * 50)

for _, row in df.iterrows():
    print(f"{row['experience']:<25} {int(row['avg_salary']):<15} руб.")

# Находим зарплату для опыта от 3 до 6 лет
experience_3_6 = df[df['experience'].str.contains('3|4|5|6', case=False, na=False)]
if not experience_3_6.empty:
    salary_3_6 = experience_3_6['avg_salary'].mean()
    print(f"\nСредняя зарплата для опыта 3-6 лет: {int(salary_3_6)} руб.")
else:
    print("\nНет данных по опыту работы 3-6 лет")

# Закрываем подключение
connection.close()
print("\nПодключение закрыто")

C:\Users\Владимир\AppData\Local\Temp\ipykernel_9656\232067519.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Средняя зарплата для DS по опыту работы:
Опыт работы               Средняя зарплата
--------------------------------------------------
Нет опыта                 74643           руб.
От 1 года до 3 лет        139675          руб.
От 3 до 6 лет             243115          руб.

Средняя зарплата для опыта 3-6 лет: 191395 руб.

Подключение закрыто


***

### выводы по предметному анализу 
Анализ вакансий показывает, что рынок data science характеризуется высоким барьером для входа — лишь около 3% предложений рассчитаны на начинающих специалистов. При этом работодатели ожидают от кандидатов широкого набора навыков: в среднем каждая вакансия требует более шести компетенций. Критически важным инструментом является Python, который требуется в каждом пятом объявлении, что подтверждает его статус обязательного языка для профессии.

Опыт работы остается ключевым фактором дохода. Зарплата в поле растет стремительно: специалисты с 3–6 годами опыта зарабатывают в среднем в 3,25 раза больше, чем начинающие. Это свидетельствует о быстрой окупаемости инвестиций в обучение и построение карьеры в данных, особенно после преодоления порога вхождения.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### 1. Общая характеристика рынка: Цифровизация и данные как новая нефть
Вывод: Российский рынок труда находится в активной фазе цифровой трансформации. Это подтверждается абсолютным лидерством IT-компаний (Яндекс, Тинькофф, СБЕР) в рейтинге работодателей. Высокий спрос на специалистов с опытом 1-3 года (26 152 вакансии) и огромное количество вакансий для новичков (7 197) говорят о том, что индустрия растет опережающими темпами и испытывает острую нехватку кадров. Компании вынуждены активно доучивать молодых специалистов внутри себя, что объясняет обилие стажерских программ.

Почему так получилось? Мировая и национальная экономика переходит на цифровые рельсы. Данные стали ключевым активом, а способность их обрабатывать — конкурентным преимуществом. Это создало взрывной спрос на всех, кто работает с данными: от аналитиков до инженеров и ученых.

Прогноз: Спрос на IT- и data-специалистов будет только расти, причем не только в pure-play IT-компаниях, но и в традиционных отраслях (ритейл, банки, телеком), которые вынуждены digitalize свои процессы, чтобы выжить.

### 2. Анализ по городам: Гиперцентрализация и зарождение спутниковых экосистем
Вывод: Рынок вакансий крайне централизован. Москва (5333) и Санкт-Петербург (2851) формируют более 16% от общего числа вакансий в базе. Это классическая модель "центр-периферия". Присутствие в топе Минска и Алматы — это не только международный охват, но и свидетельство миграции талантов и бизнеса в крупные, более стабильные центры на постсоветском пространстве.

Почему так получилось? Крупные города предлагают развитую инфраструктуру, доступ к капиталу (венчурные фонды), концентрацию университетов и пул квалифицированных специалистов. Это создает самоусиливающийся эффект: компании открываются там, где есть кадры, а кадры едут туда, где есть компании.

Что можно исследовать дополнительно?

Анализ "утечки мозгов": Сравнить долю вакансий для senior-разработчиков в регионах и в столицах. Можно ожидать, что в регионах преобладают вакансии для juniors и mids, а позиции для seniors сконцентрированы в Москве и Питере.

Поиск "хабов": Выявить города-миллионники (Новосибирск, Екатеринбург, Казань), которые показывают аномально высокий для своего размера рост IT-вакансий. Это могут быть зарождающиеся технологические хабы.

Прогноз: Централизация будет сохраняться, но рост удаленной работы (7802 вакансии) может постепенно начать менять ландшафт. У компаний из столиц появится возможность нанимать лучших специалистов из регионов, а у региональных специалистов — получать столичные зарплаты, не переезжая. Это может привести к "виртуальной децентрализации" рынка.

### 3. Анализ Data Science: Элитарность и быстрая эволюция
Вывод: Сфера Data Science из экзотической быстро превращается в mainstream-направление, но проходит через этап "взросления". Высокий барьер входа (всего 3% вакансий для новичков) и требование к широкому набору навыков (6.41 на вакансию) говорят о том, что рынок отдает предпочтение готовым специалистам-универсалам (T-shaped). Работодателям невыгодно брать новичков из-за высоких затрат на их обучение и сложности задач.

Почему зарплата растет так стремительно (в 3.25 раза за 3-6 лет)?

Высокая отдача от опыта: Опытный DS не просто "знает Python", а умеет правильно ставить задачу, выбирать модель, внедрять ее в production и считать экономический эффект. Его работа напрямую генерирует деньги (снижает затраты, увеличивает продажи).

Дисбаланс спроса и предложения: Спрос на опытных специалистов колоссален (каждая компания хочет себе DS), а предложение ограничено. Это классическая рыночная экономика: дефицитный ресурс дорожает.

Порог входа повышается: Требования к junior-ам растут. То, что было нормой для middle 5 лет назад, сейчас ждут от junior. Соответственно, и оплата для true middle/senior уровня взлетает.

Что можно исследовать дополнительно?

Анализ "навыков будущего": Не просто посчитать частоту упоминания Python/SQL, а выделить группы навыков (e.g., "MLOps" (Docker, Kubernetes), "Big Data" (Spark, Hadoop), "Deep Learning" (TensorFlow, PyTorch)) и посмотреть, в вакансиях для какого опыта они чаще встречаются. Это покажет, в какую сторону движется индустрия.

Зарплатная вилка по стекам технологий: Сравнить среднюю зарплату для вакансий, требующих, например, Python + SQL + Tableau, и вакансий, требующих Python + Spark + TensorFlow. Это покажет, какие специализации ценятся выше.

Прогноз: Барьер входа останется высоким, но сместится в сторону инженерных навыков (MLOps, облака, работа с большими данными). Pure research-позиций будет становиться меньше, а востребованность инженеров по машинному обучению (Machine Learning Engineer), которые могут не только построить модель, но и заставить ее работать в продакшене, — резко возрастет.

### Общий вывод
Российский рынок вакансий — это зеркало глобальных трендов: гиперцифровизация, централизация и война за таланты. Он разделен на два полюса: с одной стороны — динамичный, высокооплачиваемый, но труднодоступный мир IT и Data Science, сконцентрированный в столицах. С другой — традиционный рынок с доминированием полной занятости и малого бизнеса в регионах.

Ключевым драйвером изменений становится удаленный формат работы. Он является мостом между этими двумя полюсами, позволяя компаниям из центра искать таланты по всей стране и за ее пределами, а специалистам из регионов — получать доступ к лучшим проектам и зарплатам. В ближайшие годы именно этот фактор может оказать наибольшее влияние на перераспределение рынка труда, сделав его более гибким, виртуальным и менее привязанным к конкретным географическим точкам.